Add DSpace REST space

jrwdunham opened the pull request:
Connected to #391 

Replaces https://github.com/artefactual/archivematica-storage-service/pull/388

J4bbi committed:
Add DSpace REST API Space

Features:

- Can link a stored DIP to an ArchivesSpace entity, if the space
  configuration or the metadata-supplied configuration provides the
  appropriate credentials.
- Can extract DSpace and/or ArchivesSpace credentials from
  transfer-supplied metadata (JSON or CSV).

sevein commented on .travis.yml:
```diff
@@ -1,7 +1,7 @@
 
 dist: trusty
-sudo: false
+sudo: required
```
I'm hoping we don't have to go into full virtual machines because it's going to be much slower. Do you mind checking again without this change to see what's the error? Maybe it was just a temporary issue.

sevein replied:
I've confirmed this was just a temporary issue in Travis infra.

sevein commented on storage_service/locations/models/dspace_rest.py:
```diff
@@ -0,0 +1,494 @@
+from requests import RequestException
+
+# This project, alphabetical
+from agentarchives import archivesspace
```
I'm confused, this is a space that uses both ArchivesSpace and DSpace? Can you point me to a document where this integration is described?

jrwdunham replied:
I was unable to find a document for you @sevein. This development was done primarily by an Edinburgh dev (Hrafn) and, in general, they defined their own requirements internally. I have modified the DSpaceREST space creation form to make it clear that by omitting a single AS field you will be effectively "disabling" the AS functionality. @kellyannewithane or @jhsimpson may be able to provide some insight here.

sevein commented on storage_service/locations/models/dspace_rest.py:
```diff
@@ -0,0 +1,494 @@
+                headers=headers,
+                cookies={'JSESSIONID': ds_sessionid},
+                verify=self.verify_ssl)
+            LOGGER.info('Logged out of REST API.')
```
One recurrent thought I have about logging is that we should just log human-actionable events otherwise it's just noise for the application administrator. If you're a developer and you want this level of detail then use `debug`. Not something that we should start changing now but it'd be good to keep in mind, maybe someday we can apply a new strategy generally.

jrwdunham replied:
I'll cut out some of the logging cruft in any case.

sevein commented on storage_service/locations/models/dspace_rest.py:
```diff
@@ -0,0 +1,494 @@
+    verify_ssl = models.BooleanField(
+        blank=True, default=True, verbose_name=_l("Verify SSL certificates?"),
+        help_text=_l('If checked, HTTPS requests will verify the SSL'
+                     ' certificates'))
```
We also have a global `INSECURE_SKIP_VERIFY`. This is probably better or more granular but I figured I'd mention it just in case you didn't remember and it'd change your mind.

https://github.com/artefactual/archivematica-storage-service/blob/31bad497be0ce1c94f3bf3413a93b75bae9815e3/storage_service/storage_service/settings/base.py#L411-L417

jrwdunham replied:
Thanks @sevein, I had forgotten about `INSECURE_SKIP_VERIFY`. I think it's good to keep `DSpaceREST.verify_ssl` because, as you say, it is more fine-grained. This allows users to decrease their security wrt trusting responses from external services in the specific DSpace REST case but not elsewhere.

sevein commented on storage_service/locations/models/dspace_rest.py:
```diff
@@ -0,0 +1,494 @@
+    @staticmethod
+    def _debug_ds_login(url, email, password):
+        return ('using url "{url}", email "{email}", and password'
+                ' "{password}"'.format(url=url, email=email, password=password))
```
I wonder if it would be cleaner if `DSpaceRESTException` had a constructor capable of taking `(url, email, password)` and include it in the final error message. Also I'd suggest not to include the password unless the debug mode is enabled.

jrwdunham replied:
Good idea about the constructor. I've done that. Also removed the password from the exception message.

sevein commented on storage_service/locations/models/dspace_rest.py:
```diff
@@ -0,0 +1,494 @@
+            response = requests.post(collection_url,
+                                     headers=headers,
+                                     cookies={'JSESSIONID': ds_sessionid},
+                                     # data=json.dumps(item),
```
This line can be removed.

jrwdunham replied:
mais oui

sevein commented on storage_service/locations/models/dspace_rest.py:
```diff
@@ -0,0 +1,494 @@
+            msg = 'Not a JSON response'
+            LOGGER.error(msg)
+            LOGGER.error(response.text)
+            raise DSpaceRESTException(msg)
```
Do we have to log the error **and** raise and exception? It seems that the former is not needed unless you want it for debugging purposes so then I'd recommend to use the `debug` level instead, particularly when logging the body of the response.

jrwdunham replied:
Yeah, I've gone through and tried to use `raise DSpaceRESTException` consistently. Cleaner now I think.

sevein commented on storage_service/locations/models/dspace_rest.py:
```diff
@@ -0,0 +1,494 @@
+            response.raise_for_status()
+        except requests.HTTPError:
+            LOGGER.warning('Error logging in to DSpace REST API, aborting',
+                           exc_info=True)
```
This is going to include the traceback which is rather noisy.
Should we just include the message in `HTTPError`?

jrwdunham replied:
K. removed tracebacks in general.

sevein commented on storage_service/locations/models/dspace_rest.py:
```diff
@@ -0,0 +1,494 @@
+    def _deposit_aip_to_dspace(self, source_path, headers, ds_item, ds_sessionid):
+        LOGGER.info('Package is AIP. Sending file.')
+        bitstream_url = (
+            '{url.scheme}://{url.netloc}{url.path}'
```
There's more than one place where we build this base URL, should we have a dedicated function for it?

jrwdunham replied:
good idea

sevein commented on storage_service/locations/models/dspace_rest.py:
```diff
@@ -0,0 +1,494 @@
+
+        # Headers for connecting to DSpace REST API
+        headers = {'Accept': 'application/json',
+                   'Content-Type': 'application/json'}
```
Shouldn't this be a default down in the stack?

sevein reviewed (changes requested):
Nice tests @jrwdunham. Not easy because of the number of dependencies. It's the dependencies which confused me a bit. DSpace, ArchivesSpace, Tivoli... is the name of the space representative to this workflow?

I was wondering if we need a `MultiSpace` that the user can configure to combine multiple backends - perhaps a crazy idea out of scope, feel free to ignore.

jrwdunham commented:
@sevein I think I have addressed your change requests. Can you look again?

> is the name of the space representative to this workflow?

Good question. I think it is, despite the fact that DSpace, ArchivesSpace and Tivoli are involved. I have made the Tivoli upload logic configurable and disabled by default. The ArchivesSpace DIP linking stuff is closely linked to the DSpace DIP upload: the DIP is uploaded to DSpace and given a DSpace handle and that handle is provided when creating the ArchivesSpace digital object. (It may be that AIPs should also be linked/registered in AS, but that is another separate question.)

For the Tivoli stuff, the better way to do it would probably have been to create a dedicated Tivoli-specific space and set up AIP replication from the DSpace space to the Tivoli one. But at this point I think we'll have to consider that an improvement for the future.

jrwdunham committed:
Add tests for DSpace REST

Also modifies dspace_rest.py both superficially (e.g., uses format() for URL
construction) and substantively (e.g., raises exceptions more
consistently when things go wrong, makes the TSM AIP upload configurable
and disabled by default.)
